File: packages/js/core/src/types/PluginRegistration.ts
import { PluginPackage, Uri } from ".";

import { Tracer } from "@polywrap/tracing-js";
import { toUri } from "../utils";

export interface PluginRegistration<TUri extends Uri | string = string> {
  uri: TUri;
  plugin: PluginPackage<unknown>;
}

export const sanitizePluginRegistrations = Tracer.traceFunc(
  "core: sanitizePluginRegistrations",
  (input: PluginRegistration<Uri | string>[]): PluginRegistration<Uri>[] => {
    const output: PluginRegistration<Uri>[] = [];
    for (const definition of input) {
      const uri = toUri(definition.uri);

      output.push({
        uri,
        plugin: definition.plugin,
      });
    }

    return output;
  }
);

File: packages/py/core/src/types/plugin_registration.py
from __future__ import annotations
from dataclasses import dataclass
from typing import List

from . import Uri


@dataclass
class PluginRegistration:
    uri: str
    plugin: PluginPackage


def sanitize_plugin_registrations(input: List[PluginRegistration]) -> List[PluginRegistration]:
    output = []
    for definition in input:
        uri = Uri(definition.uri)
        output.append(PluginRegistration(uri=uri, plugin=definition.plugin))
    return output
